### DISCUSSION TITLE
adding this config to kubelet on RKE2 cluster

### DISCUSSION THREAD

**Discussion by thinzaung81** [rancher/rke2#7138] [Q&A] [answered]
I already have RKE2 cluster up and running already. I wanted to add this configuration `image-gc-high-threshold` to my kubelet so that it will clean up the images/continer automatically. Is there a way I can do it? Do I just add it to config.yaml file ?

Thanks

**Comment by brandond** [comment]
yep. and restart.

**Reply by nookseal** [reply]
We only do on worker node?(rke2 agent)
which file should we put and restart rke2-agent?

**Reply by brandond** [reply]
Do it on any node where you want to change that value.

**Reply by nookseal** [reply]
what should be the path we put the configuration?
I'm not seeing config.yml in /etc/rancher/rke2
but directory /etc/rancher/rke2/config.yaml.d

**Reply by brandond** [reply]
https://docs.rke2.io/install/configuration

If you provisioned via Rancher, the Rancher managed configuration is placed in a dropin file. But if you're using Rancher you should add the Kubelet args via Rancher, not by poking at the nodes directly.

Also note that the kubelet DOES have a default value for that arg that will trigger image cleanup, with you needing to set it. You should read https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/ and note the defaults.

**Reply by nookseal** [reply]
I got a bit confusing. You are saying that even we have rke2 wrapping the K8s cluster, the Kubelet on each node will contain this default value (low 80 and high 85) and performs automatically cleanup?

I wonder on doc said "DEPRECATED: This parameter should be set via the config file specified by the kubelet's --config flag. See kubelet-config-file for more information."

what does this mean actually?

`
kubelet 
--
--image-gc-high-threshold int32     Default: 85
  | The percent of disk usage after which image garbage collection is always run. Values must be within the range [0, 100], To disable image garbage collection, set to 100. (DEPRECATED: This parameter should be set via the config file specified by the kubelet's --config flag. See kubelet-config-file for more information.)
--image-gc-low-threshold int32     Default: 80
  | The percent of disk usage before which image garbage collection is never run. Lowest disk usage to garbage collect to. Values must be within the range [0, 100] and should not be larger than that of --image-gc-high-threshold. (DEPRECATED: This parameter should be set via the config file specified by the kubelet's --config flag. See kubelet-config-file for more information.)
`

**Reply by brandond** [reply]
I'm not sure what you mean by "wrapping" the cluster. RKE2 is a Kubernetes distribution. It contains all the normal Kubernetes components built directly from upstream Kubernetes project code, including the kubelet. We do not override or alter the kubelet's default image GC thresholds. You can check the rke2 service logs in journald, or look at the process list, to see what args we do start the kubelet with.

CLI flag deprecation does not mean that the flags no longer work. They still work fine, and have the documented defaults.

**Reply by nookseal** [reply]
For wrapping I refer to your message below just for reference that we use Rancher to manage kubernetes. 

"if you're using Rancher you should add the Kubelet args via Rancher, not by poking at the nodes directly"

So let say if I want to adjust some kubelet-args, image-gc-high-threshold=70 and image-gc-low-threshold=50.
I've found so many way on the other website and discussion and not sure what is the correct way.

actually there is /etc/rancher/rke2/config.yaml.d/50-rancher.yaml but as I understand we don't touch this

so we create new file in /etc/rancher/rke2/config.yaml.d/kubelet-extra-args.yaml

kubelet-arg:
  - "image-gc-high-threshold=70"     
  - "image-gc-low-threshold=50"       
  - "maximum-dead-containers=10"      
  - "maximum-dead-containers-per-container=2" 


then restart rke2-agent service on each worker

not sure if this is correct way?

**Reply by brandond** [reply]
No. You go into Rancher, edit the cluster there,  and add the desired kubelet args.

**Reply by nookseal** [reply]
I've tried looking the Rancher GUI -> CLuster Management 
try checking Edit as yaml, edit config but seem not found any option related to the extra-args for Kubernetes component related

Also, try checking the ps command for kubelet and found many parameter running on kubelet process including --cluster-dns, --eviction-hard, --eviction-minimum-reclaim 

but not found parameter with image-gc low and high on the ps command output, normally it's run as default if not specify anything?

will try checking more.

**Reply by brandond** [reply, marked as answer]
Where exactly were you looking in the Rancher UI?
![image](https://github.com/user-attachments/assets/e4f5463f-5a8c-4dbd-b534-4d4f07125402)

**Reply by nookseal** [reply]
Thank you, Follow your screenshot found this on rke2, do we have this option available on rke1?

**Reply by dereknola** [reply]
No idea on if its available to RKE1. FYI RKE1[ EOL has been announced](https://www.suse.com/support/kb/doc/?id=000021513).

**Reply by nookseal** [reply]
Thanks for your kind support both
